File: 2-Add-Two-Numbers.py
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode()
        curr = dummy 
        carry = 0
        # Testing            
        while l1 or l2 or carry:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0

            total = num1 + num2 + carry
            carry = total // 10

            curr.next = ListNode(total % 10)
            curr = curr.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return dummy.next

File: neetcode/2-Add-Two-Numbers.ts
function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
  const num1: number[] = [];
  const num2: number[] = [];

  while (l1) {
    num1.unshift(l1.val);
    l1 = l1.next;
  }
  while (l2) {
    num2.unshift(l2.val);
    l2 = l2.next;
  }

  const sum = BigInt(num1.join("")) + BigInt(num2.join(""));
  const digits = sum.toString().split("").reverse();

  const dummy = new ListNode(0);
  let curr = dummy;

  for (const digit of digits) {
    curr.next = new ListNode(Number(digit));
    curr = curr.next;
  }

  return dummy.next;
}